xchange-kraken/src/main/java/info/bitrich/xchangestream/kraken/KrakenStreamingTradeService.java

```diff
@@ -0,0 +1,221 @@
+import org.knowm.xchange.dto.trade.StopOrder;
+import org.knowm.xchange.dto.trade.UserTrade;
+import org.knowm.xchange.kraken.KrakenAdapters;
+import org.knowm.xchange.kraken.dto.account.KrakenWebsocketToken;
+import org.knowm.xchange.kraken.dto.trade.KrakenOrderFlags;
+import org.knowm.xchange.kraken.dto.trade.KrakenOrderStatus;
+import org.knowm.xchange.kraken.dto.trade.KrakenType;
+import org.knowm.xchange.kraken.service.KrakenAccountServiceRaw;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+public class KrakenStreamingTradeService implements StreamingTradeService {
+    private static final Logger LOG = LoggerFactory.getLogger(KrakenStreamingTradeService.class);
+
+    private KrakenAccountServiceRaw rawKrakenAcctService;
+    private KrakenStreamingService streamingService;
+    private KrakenWebsocketToken token;
+    private long tokenExpires = 0L;
+
+    KrakenStreamingTradeService(KrakenStreamingService streamingService, KrakenAccountServiceRaw rawKrakenAcctService) {
+        this.streamingService = streamingService;
+        this.rawKrakenAcctService = rawKrakenAcctService;
+    }
+    private String getChannelName(KrakenSubscriptionName subscriptionName) {
+        return subscriptionName.toString();
+    }
+
+    private static class KrakenDtoOrderHolder extends HashMap<String, KrakenOpenOrder> {}
+    private static class KrakenDtoUserTradeHolder extends HashMap<String, KrakenOwnTrade> {}
+
+    private KrakenWebsocketToken renewToken() throws IOException {
+        if (System.currentTimeMillis() >= tokenExpires) {
+            token = rawKrakenAcctService.getKrakenWebsocketToken();
+            tokenExpires = System.currentTimeMillis() + (token.getExpiresInSeconds() * 900L); // 900L instead of a 1000L for 90%
+        }
+        return token;
+    }
+
+    @Override
+    public Observable<Order> getOrderChanges(CurrencyPair currencyPair, Object... args) {
+
+        return Observable.create( source -> {
+            try {
+                String channelName = getChannelName(KrakenSubscriptionName.openOrders);
+
+                streamingService.subscribeChannel(channelName, renewToken().getToken())
+                        .filter(JsonNode::isArray)
+                        .filter(Objects::nonNull)
+                        .map(jsonNode -> jsonNode.get(0))
+                        .map(jsonNode ->
+                                StreamingObjectMapperHelper.getObjectMapper().treeToValue(jsonNode, KrakenDtoOrderHolder[].class))
+                        .doOnError(source::onError)
+                        .forEach(list ->
+                                adaptKrakenOrders(list)
+                                        .forEach(source::onNext));
+
+            } catch (Throwable e) {
+                source.onError(e);
+            }
+        });
+    }
+
+    private List<Order> adaptKrakenOrders(KrakenDtoOrderHolder[] dtoList) {
```

Comment:
If you changed this method to return `Iterable<Order>` you could do the conversion using an Guava `FluentIterable` and avoid creating a temporary list. Minor suggestion, not required to pass review.

Reply:
no objection, it does not break anything.